cpp/src/arrow/compute/kernels/vector_sort_test.cc

```diff
@@ -364,32 +400,264 @@ TYPED_TEST(TestSortToIndicesKernelRandomCount, SortRandomValuesCount) {
+      for (auto order : {SortOrder::ASCENDING, SortOrder::DESCENDING}) {
+        auto array = rand.Generate(length, null_probability);
+        ASSERT_OK_AND_ASSIGN(std::shared_ptr<Array> offsets, SortIndices(*array, order));
+        ValidateSorted<ArrayType>(*checked_pointer_cast<ArrayType>(array),
+                                  *checked_pointer_cast<UInt64Array>(offsets), order);
+      }
     }
   }
 }
 
+class TestTableSortIndices : public ::testing::Test {
+ protected:
+  void AssertSortIndices(const std::shared_ptr<Table> table, const SortOptions& options,
+                         const std::shared_ptr<Array> expected) {
+    ASSERT_OK_AND_ASSIGN(auto actual, SortIndices(*table, options));
+    AssertArraysEqual(*expected, *actual);
+  }
+
+  void AssertSortIndices(const std::shared_ptr<Table> table, const SortOptions& options,
+                         const std::string expected) {
+    AssertSortIndices(table, options, ArrayFromJSON(uint64(), expected));
+  }
+};
+
+TEST_F(TestTableSortIndices, SortNull) {
+  auto table = TableFromJSON(schema({
+                                 {field("a", uint8())},
+                                 {field("b", uint8())},
+                             }),
+                             {"["
+                              "{\"a\": null, \"b\": 5},"
+                              "{\"a\": 1,    \"b\": 3},"
+                              "{\"a\": 3,    \"b\": null},"
+                              "{\"a\": null, \"b\": null},"
+                              "{\"a\": 2,    \"b\": 5},"
+                              "{\"a\": 1,    \"b\": 5}"
+                              "]"});
+  SortOptions options(
+      {SortKey("a", SortOrder::ASCENDING), SortKey("b", SortOrder::DESCENDING)});
+  this->AssertSortIndices(table, options, "[5, 1, 4, 2, 0, 3]");
+}
+
+TEST_F(TestTableSortIndices, SortNaN) {
+  auto table = TableFromJSON(schema({
+                                 {field("a", float32())},
+                                 {field("b", float32())},
+                             }),
+                             {"["
+                              "{\"a\": null, \"b\": 5},"
+                              "{\"a\": 1,    \"b\": 3},"
+                              "{\"a\": 3,    \"b\": null},"
+                              "{\"a\": null, \"b\": null},"
+                              "{\"a\": NaN,  \"b\": null},"
+                              "{\"a\": NaN,  \"b\": 5},"
+                              "{\"a\": NaN,  \"b\": NaN},"
+                              "{\"a\": 1,    \"b\": 5}"
+                              "]"});
+  SortOptions options(
+      {SortKey("a", SortOrder::ASCENDING), SortKey("b", SortOrder::DESCENDING)});
+  this->AssertSortIndices(table, options, "[7, 1, 2, 5, 6, 4, 0, 3]");
+}
+
+using RandomParam = std::tuple<std::string, double>;
+class TestTableSortIndicesRandom : public testing::TestWithParam<RandomParam> {
+  class Comparator : public TypeVisitor {
```

Comment:
Can you add comments when you define non-trivial classes and methods? This would help reading the code later.

Reply:
Added.